exchange/cachetest/defaultvideo.json

```diff
@@ -27,16 +27,22 @@
     "expectedCacheables": [
         {
             "Type": "json",
-            "TTLSeconds": 660
+            "TTLSeconds": 660,
+            "Data":"\"\u003cVAST version=\"3.0\"\u003e\u003cAd\u003e\u003cWrapper\u003e\u003cAdSystem\u003eprebid.org wrapper\u003c/AdSystem\u003e\u003cVASTAdTagURI\u003e\u003c![CDATA[]]\u003e\u003c/VASTAdTagURI\u003e\u003cImpression\u003e\u003c/Impression\u003e\u003cCreatives\u003e\u003c/Creatives\u003e\u003c/Wrapper\u003e\u003c/Ad\u003e\u003c/VAST\u003e\"" 
         }, {
             "Type": "json",
-            "TTLSeconds": 3660
+            "TTLSeconds": 3660,
+            "Data": "\"\u003cVAST version=\"3.0\"\u003e\u003cAd\u003e\u003cWrapper\u003e\u003cAdSystem\u003eprebid.org wrapper\u003c/AdSystem\u003e\u003cVASTAdTagURI\u003e\u003c![CDATA[]]\u003e\u003c/VASTAdTagURI\u003e\u003cImpression\u003e\u003c/Impression\u003e\u003cCreatives\u003e\u003c/Creatives\u003e\u003c/Wrapper\u003e\u003c/Ad\u003e\u003c/VAST\u003e\"" 
```

Comment:
We are currently caching Json encoded data of the bids (winning bids only from now on because of the modifications made by this PR). The `prebid_cache_client.Cacheable` struct we currently use to cache bids includes this Data field were we marshall bid data such as cat, imp, price and so on. This data field was included in the expected cachables of the test Json files in order to get compared to what we actually cache.

Reply:
Why are the values identical for the two different bids?

Reply:
I copied what the system cached. In all honesty I don understand what the banner request caches, but not what the video is caching. In the case of the banners I wrote the Data field of the expected output myself but not in this file that tests for video cache.

Reply:
So it seems there are multiple issues here. 
 
First problem is that the last two lines of this file, as well as those in all the other files in cachetest/, are wrong.  In the old auction_test.go code you can see that _TestCacheJSON_() calls _runCacheSpec(t, fileDisplayName, specData, true, false)_.  

Secondly, it should be intuitive that two different bids should not have identical values cached.  Again, if you look at the old code, you'll see that the data portion of cached vast bids was never tested.  Since this should not be vast (perhaps you already can figure it out since you understand banner request caches), the expected data is the marshal'd bid.  

Third, now that I understand that the "Data" field you've added to the JSON files is simply the marshal'd bid, I believe adding the "Data" fields, which are simply marshal'd versions of bid data already present is not a good idea.  There is a potential problem that some future JSON marshal() code will change the order of the fields.  In that case, the old approach would be okay because the same marshaling procedure will be applied to both the test data and the real data, while this new approach will cause failures because only the real data will use the new marshaling.  I believe you should undo that change and go back to the original implementation that did the marshal operation in real time, i.e.:
                cjson, _ := json.Marshal(bid.bid)
                specData.ExpectedCacheables[i].Data = cjson

Reply:
_Joe, answers shown below:_
**1)** The last two lines of every json file end up becoming the last two boolean values of the old `runCacheSpec(t, fileDisplayName, specData, true, false)` function because its signature ended up changing to `runCacheSpec(t *testing.T, fileDisplayName string, specData *cacheSpec)` in this new version the last two values found in the json files, get assigned to the target data object in lines 182 and 183 below:
```
118 func runCacheSpec(t *testing.T, fileDisplayName string, specData *cacheSpec) {
119     var bid *pbsOrtbBid
120     winningBidsByImp := make(map[string]*pbsOrtbBid)
121 +-- 38 lines: winningBidsByBidder := make(map[string]map[openrtb_ext.BidderName]*pbsOrtbBid)-------------------
159     targData := &targetData{
160         priceGranularity: openrtb_ext.PriceGranularity{
161             Precision: 2,
162             Ranges: []openrtb_ext.GranularityRange{
163                 {
164                     Min:       0,
165                     Max:       5,
166                     Increment: 0.05,
167                 },
168                 {
169                     Min:       5,
170                     Max:       10,
171                     Increment: 0.1,
172                 },
173                 {
174                     Min:       10,
175                     Max:       20,
176                     Increment: 0.5,
177                 },
178             },
179         },
180         includeWinners:    specData.TargetDataIncludeWinners,
181         includeBidderKeys: specData.TargetDataIncludeBidderKeys,
182         includeCacheBids:  specData.TargetDataIncludeCacheBids,
183         includeCacheVast:  specData.TargetDataIncludeCacheVast,
184     }
185 
```

Reply:
**2)** Excepting for the Data field, the video tests below don't cache the exact same data, se are comparing the cache in two parts:
  + The first part is the Data and the TTLseconds fields in lines 207 and 222 of auction _test.go
  + If the first part matches and we are testing a video request, we compare a second part of the cache, the keys. This is performed in the true branch of the else statment in line 239 in auction_test.go 
```
117 // finds the highest bid of every Imp.
118 func runCacheSpec(t *testing.T, fileDisplayName string, specData *cacheSpec) {
119 +--115 lines: var bid *pbsOrtbBid-------------------------------------------------------------------------------------
234         for k, cachedElements := range ht {
235             if cachedElements.freq > 0 {
236                 t.Errorf("%s:  [CACHE_ERROR] Cache inconsistency. Element %s was not expected to get cached\n", fileDisplayName, k)
237             } else if cachedElements.freq < 0 {
238                 t.Errorf("%s:  [CACHE_ERROR] Cache inconsistency. We cached some more elements (%s) than expected.\n", fileDisplayName, k)
239             } else if targData.includeCacheVast {
240                 for i := 0; i < len(cachedElements.expectedKeys); i++ {
241                     found := false
242                     for j := 0; j < len(cachedElements.actualKeys); j++ {
243                         if strings.HasPrefix(cachedElements.expectedKeys[i], cachedElements.actualKeys[j]) {
244                             found = true
245                         }
246                     }
247                     if !found {
248                         t.Errorf("[CACHE_ERROR] Key \"%s\" was expected to be cached but was not.\n", cachedElements.expectedKeys[i])
249                     }
250                 }
251             }
252         }
253     }
254 }
```

Reply:
**3)** That's not exactly a good idea because the data doesn't come cached in the exact same order than listed in the json file. If we just marshal the fields as they come in the json file and try to compare side by side, we'll get something like the following:
```
              Expected 2                                                                      Actual 2
Type      | json                                                                  | json
Data      | "{ \"id\": \"bidTwo\", \"impid\": \"oneImp\", \"price\": 5.64 }" | {"id":"bidOne","impid":"oneImp","price":7.64,"exp":600} 
TTLSeconds| 360                                                                   | 660
Key       | ""                                                                    | "" 
          |                                                                       |   
Type      | json                                                                  | json
Data      | "{ \"id\": \"bidOne\", \"impid\": \"oneImp\", \"price\": 7.64, \"exp\": 600}" | {"id":"bidTwo","impid":"oneImp","price":5.64} 
TTLSeconds| 660                                                                   | 360
Key       | ""                                                                    | "" 
          |                                                                       |   
```
The data is clearly there but it wasn't cached in the exact same order. If we do a direct comparison of the 5.64 bid vs the 7.64 bid they won't match. And if we want to go find a match of the first element of the left in the entire "Actual" list on the right, we'll have an O(n^2) process that is quite inefficient. The O(n+m) hashing approach that we coded is far better specially if we increase the size of the tests in the future. Therefore, I thought the approach of listing in the expected cachables part of the json file what you expect to assert in the test output, was better.

Reply:
(1)  Understood, but you've entered the wrong values on the last two lines.

(2) This is really tangential since the real problem is due to the wrong entries on the last two lines.  However, you should be aware that PBS caches only the data.  The key values, when present in the tests, represent the custom cache keys that are created by doCache() in certain cases to be used for retrieving the cached data.  See the useCustomCacheKey variable in doCache() if you want to understand how/when that occurs.  Since you've forced me to look a little further, I now see that the reason the Data values are identical for different bids is probably due to the fact that the NURL field of the bids must be either empty or identical.  If you look at doCache(), you'll see in the "bids" case, the Data value comes from
 **json.Marshal(topBidPerBidder.bid)** 
so values for different bids will always be different, while in the "vast" case, it comes from:
**vast := makeVAST(topBidPerBidder.bid)**
**jsonBytes, err := json.Marshal(vast)**
where makeVast() is simply:
func makeVAST(bid *openrtb.Bid) string {
        if bid.AdM == "" {
                return `<VAST version="3.0"><Ad><Wrapper>` +
                        `<AdSystem>prebid.org wrapper</AdSystem>` +
                        `<VASTAdTagURI><![CDATA[` + bid.NURL + `]]></VASTAdTagURI>` +
                        `<Impression></Impression><Creatives></Creatives>` +
                        `</Wrapper></Ad></VAST>`
        }
        return bid.AdM
}
You can see that if bid.NURL is empty or identical for both bids, the Data values cached will be identical.  Presumably, that means that bid.NURL must exist and be unique in real (non-test) cases (though I suppose it could mean we have a flawed caching approach, too).  Thus, if you want to test the Data values for vast bids, you simply need to specify a unique bid.NURL value for each bid and then confirm the presence of what would be created from the above calls.  Or you could go back to skipping checking the Data values for vast bids, as the old test code did.

(3) You may be right, in which case both approaches are flawed.  If so, the correct approach is to unmarshal both sets of data into appropriate structures and compare the corresponding fields of interest in the two structures.  BTW, the performance characteristics here are pretty irrelevant given that the numbers will always be quite small and that these are only tests.